File: dan/Problems/Hard/Graph/753. Cracking the Safe/solution.java
class Solution {
    StringBuilder res = new StringBuilder();
    Set<String> set = new HashSet<>();
    
    public String crackSafe(int n, int k) {
        StringBuilder sb = new StringBuilder();
        for(int i = 1; i < n; i++){
            sb.append("0");
        }
        dfs(sb.toString(), k);
        for(int i = 1; i < n; i++){
            res.append("0");
        }
        return res.toString();
        
    }
    
    public void dfs(String node,int k){
        for(int i = 0; i < k; i++){
            String psw = node.concat(String.valueOf(i));
            if(!set.contains(psw)){
                set.add(psw);
                dfs(psw.substring(1), k);
                res.append(i);
            }
        }
    }
}

File: dan/Problems/Hard/Graph/753. Cracking the Safe/solution.py
class Solution(object):
    def crackSafe(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: str
        """
        seen = set()
        ans = []
        def dfs(node):
            for i in map(str, range(k)):
                pos = node + i
                if not pos in seen:
                    seen.add(pos)
                    dfs(pos[1:])
                    ans.append(i)
        dfs("0" * (n - 1))
        return "".join(ans) + "0" * (n - 1)
        
